# Prokaryote isolate sequencing: Data processing and assembly (Oxford Nanopore)

----

### Data processing and assembly of prokaryote culture isolates based on Oxford Nanopore long-read sequencing data.

Workflow for processing raw prokaryote isolate Nanopore sequencing data through to assembly.

The workflow outlined here includes:

- Nanopore data qc, trimming, and filtering
- Assembly

This workflow was developed within the group of Associate Professor Kim Handley (University of Auckland) under the Environmental Metagenomics branch of Genomics Aotearoa. 

It is intended as companion document to:

* downstream processing workflows including: Prokaryote_genomics; Viromics_WGS
* the metagenomics summer school materials

Scripts and software usage are directed towards researchers based in Aotearoa New Zealand, and who have access to the *New Zealand eScience Infrastructure* (NeSI). Installation and maintainance of the relevant software is not discussed here, and we are not resourced to support adapting this workflow to other settings. However, for users who have appropriate computing resources available with the necessary software installed, this information should be entirely transferrable.

**Dr Michael Hoggard**<BR>
The University of Auckland <BR>
Genomics Aotearoa<BR>
February 2023

----
