{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e139318f-4e2a-491b-8cac-d187e8811749",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Taxonomy and gene-sharing network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b51bb9b-02f2-408c-adb5-0e4961c4a734",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95c12766-950c-4306-abdd-1f66136ac0d3",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "Viral taxonomy in general is a challenging and always evolving field. For example, 2022 saw a large overhaul of taxonomic idetification and ranking structures. For more information on viral taxonomy, see the website for the International Committee on Taxonomy of Viruses: https://ictv.global/\n",
    "\n",
    "In addition to this, predicting viral taxonomy remains challenging in metagenomics-based studies.  \n",
    "\n",
    "One leading approach currently involves generating a gene-sharing network of the similarity of all viral genomes (or contigs) compared to reference genomes based on the extent of shared genes. \n",
    "\n",
    "*vConTACT2* can generate a gene sharing network including the viral-RefSeq reference database, and which can be used to infer the potential taxonomy of your data set's viral contigs. *vConTACT2* generates this output based on predicted genes (e.g. *prodigal* output file). Taxonomy can then be *inferred* (note, this is note a definitive taxonomy assignment) based on clustering with reference genomes. \n",
    "\n",
    "Key outputs from *vConTACT2* are:\n",
    "\n",
    "- `genome_by_genome_overview.csv`: clustering results of all included contigs (also including ViralRefSeq genomes). In some cases, taxanomy can be inferred based on how closely your query viral genome clusters with a reference genome(s). The script `tax_predict.py` was written to automate taxonomy predictions *en masse* based on this output file. However, note that in practice this results in only a very small percentage of taxonomy predictions overall (from metagenome-based data sets). It is generally necessary to examine the clustering statistics and visualise the network to assess the likely taxonomy of each viral genome of interest.\n",
    "- `c1.ntw`: network file that can be visualised via, e.g. *cytoscape*\n",
    "\n",
    "Further information on *vConTACT2* can be found here: https://bitbucket.org/MAVERICLab/vcontact2/src/master/\n",
    "\n",
    "A helpful example (including visualisation in cytoscape) is available here: https://www.protocols.io/view/applying-vcontact-to-viral-sequences-and-visualizi-kqdg3pnql25z/v5?step=7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55ae8c46-d39c-4c15-9b66-5c99740c2c34",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa48d589-4fca-49aa-ab91-0f70d4623895",
   "metadata": {},
   "source": [
    "## Run vConTACT2\n",
    "\n",
    "*vConTACT2* is not currently available in NeSI and must be installed as a conda environment (Note, you may also need to install `cluster_one-1.0.jar` separately). It is advisable to discuss conda installs with a member of the NeSI support team. \n",
    "\n",
    "The example below assumes a conda install is available, and runs all required steps in one slurm script: *prodigal* gene prediction; generates *vConTACT2* gene2genome lookup table; running *vConTACT2*; running tax_predict.py\n",
    "\n",
    "NOTE:\n",
    "\n",
    "- In this example, we are running *prodigal* to first predict all genes in our viral contig data set. It may be possible to simply take this directly from the *DRAM-v* output (assuming the formatting is appropriate for *vConTACT2*)\n",
    "- Run `vcontact2 -h` to see the available RefSeq reference databases\n",
    "- The version of `tax_predict.py` below was based on the output generated by vConTACT2.0.11.3, and may not be functional with other versions of vConTACT2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4119ba49-d08e-4aff-8d3c-a1048284e6d9",
   "metadata": {},
   "source": [
    "#### If required: Install ClusterONE into conda/envs/vContact2/bin/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aae718aa-5a69-4966-acc5-c616744097d2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Install ClusterONE\n",
    "cd /path/to/conda/envs/vContact2/bin/\n",
    "wget https://paccanarolab.org/static_content/clusterone/cluster_one-1.0.jar"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e39e7172-078f-4ac5-b0fe-f438c4337b86",
   "metadata": {},
   "source": [
    "#### Run vConTACT2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c44af47b-9b9e-4ee4-af81-6517dbbaaa87",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A your_project_account\n",
    "#SBATCH -J 6_Taxonomy_vcontact2\n",
    "#SBATCH --time 03:00:00\n",
    "#SBATCH --mem=20GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=32\n",
    "#SBATCH -e 6_Taxonomy_vcontact2.err\n",
    "#SBATCH -o 6_Taxonomy_vcontact2.out\n",
    "#SBATCH --profile=task\n",
    "\n",
    "# Set up working directories\n",
    "cd /working/dir\n",
    "mkdir -p 4.taxonomy/vConTACT2\n",
    "\n",
    "### Prodigal gene prediction\n",
    "\n",
    "# Load dependencies\n",
    "module purge\n",
    "module load Prodigal/2.6.3-GCC-9.2.0\n",
    "\n",
    "# Run prodigal\n",
    "srun prodigal -p meta -q \\\n",
    "-i 1.viral_identification/6.checkv_vOTUs/vOTUs.checkv_filtered.fna \\\n",
    "-a 4.taxonomy/vConTACT2/vOTUs.faa \n",
    "\n",
    "### vConTACT2 gene2genome\n",
    "\n",
    "cd /working/dir\n",
    "\n",
    "# Generate gene2genome mapping file for vConTACT2\n",
    "module purge\n",
    "module load Miniconda3\n",
    "source activate vContact2\n",
    "# Binary paths\n",
    "export PATH=\"/path/to/conda/envs/vContact2/bin:$PATH\"\n",
    "# Dependency module loads\n",
    "module load DIAMOND/0.9.32-GCC-9.2.0\n",
    "module load MCL/14.137-gimkl-2020a\n",
    "\n",
    "# run with prodigal output .faa file\n",
    "vcontact2_gene2genome -p 4.taxonomy/vConTACT2/vOTUs.faa -o 4.taxonomy/vConTACT2/viral_genomes_g2g.csv -s 'Prodigal-FAA'\n",
    "\n",
    "### Run vConTACT2\n",
    "\n",
    "# Run vcontact2\n",
    "srun vcontact2 \\\n",
    "-t 32 \\\n",
    "--raw-proteins 4.taxonomy/vConTACT2/vOTUs.faa \\\n",
    "--rel-mode Diamond \\\n",
    "--proteins-fp 4.taxonomy/vConTACT2/viral_genomes_g2g.csv \\\n",
    "--db 'ProkaryoticViralRefSeq211-Merged' \\\n",
    "--c1-bin /path/to/conda/envs/vContact2/bin/cluster_one-1.0.jar \\\n",
    "--output-dir 4.taxonomy/vConTACT2/vConTACT2_Results\n",
    "\n",
    "### deactivate conda environment\n",
    "\n",
    "conda deactivate\n",
    "\n",
    "### tax_predict.py\n",
    "\n",
    "# Load dependencies\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "\n",
    "# Set up working directories\n",
    "cd /working/dir\n",
    "\n",
    "# Run tax_predict.py\n",
    "/path/to/scripts/tax_predict_vConTACT2.0.11.3.py \\\n",
    "-i 4.taxonomy/vConTACT2/vConTACT2_Results/genome_by_genome_overview.csv \\\n",
    "-o 4.taxonomy/vConTACT2/vOTUs_\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69a50038-d22b-465b-8f66-0b14d9018cc5",
   "metadata": {},
   "source": [
    "***"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ],
   "version": "0.23.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
