# Environmental Metagenomics: Data processing (Metatranscriptomics (Illumina HiSeq))

----

### Data processing based on whole transciptome sequencing (WTS) data generated via Illumina HiSeq

WiP...


### Placeholder info

Workflow for processing raw whole transciptome sequencing (WTS) data (Illumina HiSeq).

- WTS Illumina HiSeq data qc, trimming, and filtering

This workflow was developed within the group of Associate Professor Kim Handley (University of Auckland) under the Environmental Metagenomics branch of Genomics Aotearoa. 

It is intended as companion document to:

* downstream processing workflows including: Viromics_WTS (WiP), and Prokaryote_genomics and Viromics_WGS (e.g. read mapping to generate coverage profiles of transcription across samples)
* the metagenomics summer school materials 

Scripts and software usage are directed towards researchers based in Aotearoa New Zealand, and who have access to the *New Zealand eScience Infrastructure* (NeSI). Installation and maintainance of the relevant software is not discussed here, and we are not resourced to support adapting this workflow to other settings. However, for users who have appropriate computing resources available with the necessary software installed, this information should be entirely transferrable.

**Dr Michael Hoggard**<BR>
The University of Auckland <BR>
Genomics Aotearoa<BR>
Februrary 2023

----
